### ISSUE TITLE
Feature-Request: Policy plugin-system for server-side event validation

### ISSUE THREAD

**Issue by OssWhisperer** [EventStore/EventStore#2945]
**Is your feature request related to a problem? Please describe.**
As a developer with sql database background (as many others), i would expect some equivalent to CHECK and UNIQUE constraints from a database engine.

**Describe the solution you'd like**
Please provide some policy system that enables service-side custom code to validate and modify events before they are added to a stream. I'd prefer some a C# interface to implement that logic and a REST-API for upload/update of the plugin/policy-assembly and dependent files in a CI/CD pipeline.

As a second step, some opt-in standard policies / implementations could be provided e.g. to:
- validate event data against json schema
- validate event-type is expected in stream-category
- validate event-type is in expected order (Archived-Event only after Created-Event)
- use JSONPath to extract a value to check for uniqueness or against regex
- insert a random-unqiue or auto-increment integer value to json event data

The custom policy code may use other streams, e.g. starting with "$_#plugin-name#", to manage their behaviour e.g. store used unique values and build some in-memory read-model for fast validation.

Of course, that policy/plugin development will require good code for performance, stability, security, scalability, failover etc. That's why there should be some standard implementations provided. A policy execution will have a performance-penalty, but impact on scalability of overall EventStoreDB should be low (or nothing at all).

If requirements are more advanced e.g. binary event data or more sophisticated unique-checks, the standard implementations could be used as starting point for own policy validation code.

**Describe alternatives you've considered**
Using json schema to validate events before inserting to database is no big deal. Since inserted events are more or less not deletable, it would be nice to have some enforcement.

I am not aware of an easy-to-use standard way to get something like auto-incement or check for uniqueness. If there is, i am happy to use it. I've only seen (and used) concepts that involve a lot of programming (like "use a stream to store the unique value first with expected-version / optimistic concurrency and then use that value") or encoding the unique-value in the stream-name (which makes changing the mail address of a user-account a bit difficult). However, with cleanup-code, retry etc. this gets a nice project on itself and may not be acceptable for small project as other DB's require less effort. I believe it's one of the args against event sourcing when trying out that concept/tech.

**Additional context**
Event metadata is already used by EventStoreDB to store $causationId and $correlationId to allow diagnostics and analysis. Event metadata could also be used to provide information about required policy or unique key usage in the event. E.g. freeing a unqiue value that was used by a previous event like UserAccountMailAddressChanged => old value can be used again, new value must be unqiue.

You get very fast a lot of ideas that would fill some more chapters, however, i tried to summarize it to get the idea, not the solution ...

**Comment by alexeyzimarev** [comment]
Sounds like schema registry. We have something similar in plans, but it won't be in this repo.
